RCode/NEFI/partition_uncertainty_example-RandWalk-Year.R

```diff
@@ -0,0 +1,247 @@
+
+#5) Save and Process Output
+
+plot(jags.out, vars = jags_plug_ins$variable.names.model) 
+
+jags.out.mcmc <- as.mcmc.list(jags.out)
+out <- as.matrix(jags.out.mcmc)
+
+DIC=dic.samples(j.model, n.iter=5000)
+DIC
+
+
+#6) CI, PI, Obs PI Calculations
+
+times <- as.Date(as.character(dat$cal$date))
+time.rng = c(1,length(times)) ## adjust to zoom in and out
+ciEnvelope <- function(x,ylo,yhi,...){
+  polygon(cbind(c(x, rev(x), x[1]), c(ylo, rev(yhi),
+                                      ylo[1])), border = NA,...) 
+}
+
+mus=grep("mu", colnames(out))
+mu = exp(out[,mus])
+ci <- apply(exp(out[,mus]),2,quantile,c(0.025,0.5,0.975))
+preds_plug_ins <- preds_plug_ins(model_name = model_name)
+pi <- apply(exp(preds_plug_ins$pred.model),2,quantile,c(0.025,0.5,0.975), na.rm=TRUE)
+obs_pi <- apply(preds_plug_ins$pred_obs.model,2,quantile,c(0.025,0.5,0.975), na.rm=TRUE)
+
+
+#7) CI, PI, Obs PI Plots
+
+#Obs vs. Latent
+plot(times,ci[2,],type='n',ylim=range(y+.01,na.rm=TRUE), log="y", ylab="observed Gloeo colonies",xlim=times[time.rng], main="Obs, Latent CI")
+ciEnvelope(times,ci[1,],ci[3,],col="lightBlue")
+points(times,y,pch="+",cex=0.5)
+
+#CI, PI, Obs PI
+plot(times,ci[2,],type='n',ylim=range(y+.01,na.rm=TRUE), log = "y", ylab="Gloeo count",xlim=times[time.rng], main="Obs, Latent CI (blue), PI (green), Obs PI (grey)")
+ciEnvelope(times,obs_pi[1,]+0.0001,obs_pi[3,],col="gray")
+ciEnvelope(times,pi[1,],pi[3,],col="Green")
+ciEnvelope(times,ci[1,],ci[3,],col="lightBlue")
+points(times,y,pch="+",cex=0.5)
+
+
+## Forward Simulation
+
+### settings
+Nmc = 1000         ## set number of Monte Carlo draws
+N.cols <- c("red","green","blue","orange") ## set colors
+ylim = range(ci)
+trans <- 0.8       ## set transparancy
+cal_time = as.Date(as.character(dat$cal$date))
+forecast_time = as.Date(as.character(dat$forecast$date))
+all_time = c(cal_time, forecast_time)
+
+plot.run <- function(){
+  plot(all_time, all_time, type='n', ylim=ylim, ylab="Gloeo count", log = 'y')
+  ecoforecastR::ciEnvelope(cal_time, ci[1,], ci[3,], col="lightBlue")
+  lines(cal_time, ci[2,], col="blue")
+  points(cal_time, ci[2,])
+}
+
+ci <- apply(exp(out[ ,mus]), 2, quantile, c(0.025,0.5,0.975))
+plot.run()
+
+
+
+#setting up deterministic forecast function for random walk with random year
+# IC = initial conditions from end of calibration period 
+# N_out = number of time steps to forecast 
+# tau_add = process error (defaults to zero) 
+# tau_yr = random effect error? (also defaults to zero?)
+# n = number of monte carlo sims 
+forecast_gloeo <- function(IC, N_out = N_out, tau_add = 0, tau_yr = 0, n = Nmc){
+  out <- matrix(NA, n, N_out) # setting up output 
+  gloeo_prev <- IC
+  for(i in 1:N_out){
+    out[,i] <- rnorm(n, gloeo_prev, tau_add) # predict next step JB getting error here when include tau_yr - unsure why?
+    gloeo_prev <- out[,i] # update IC 
+  }
+  return(out)
+}
+
+## deterministic predictions 
+N_out = nrow(dat$forecast) # length of forecast time points 
+IC = mu
+
```

Comment:
```suggestion
years_mean = apply(out[,grep('yr[',colnames(out), fixed = T)], 2, mean) # mean of the random effect for each year 
```
Calculate the mean of each year's random effect. When we forecast, we'll be using the last year's mean random effect